tenant/pkg/controller/tenant/tenant_controller.go

```diff
@@ -95,5 +98,47 @@ func (r *ReconcileTenant) Reconcile(request reconcile.Request) (reconcile.Result
 		return reconcile.Result{}, err
 	}
 
+	if instance.Spec.TenantAdminNamespaceName != "" {
+		nsList := &corev1.NamespaceList{}
+		err := r.List(context.TODO(), &client.ListOptions{}, nsList)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		expectedOwnerRef := metav1.OwnerReference{
+			APIVersion: tenancyv1alpha1.SchemeGroupVersion.String(),
+			Kind:       "Tenant",
+			Name:       instance.Name,
+			UID:        instance.UID,
+		}
+		for _, each := range nsList.Items {
+			if each.Name == instance.Spec.TenantAdminNamespaceName {
+				// Check OwnerReference
+				found := false
+				for _, ownerRef := range each.OwnerReferences {
+					if ownerRef == expectedOwnerRef {
+						found = true
+						break
+					}
+				}
+				if !found {
+					err = fmt.Errorf("TenantAdminNamespace %v is owned by %v", each.Name, each.OwnerReferences)
+				}
+				return reconcile.Result{}, err
+			}
+		}
+		tenantAdminNs := &corev1.Namespace{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: corev1.SchemeGroupVersion.String(),
+				Kind:       "Namespace",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            instance.Spec.TenantAdminNamespaceName,
+				OwnerReferences: []metav1.OwnerReference{expectedOwnerRef},
+			},
+		}
+		if err := r.Client.Create(context.TODO(), tenantAdminNs); err != nil {
+			return reconcile.Result{}, err
```

Comment:
Wouldn't we want to return a requeue here so it attempts the create again on error? I've seen a few operators that don't properly requeue cause issues when the api server is in a bad state and returns an error during a create due simply to a network blip or similar. Note: I am not an expert on kubebuilder.

```
reuturn reconcile.Result{
        RequeueAfter: time.Second, // probably want some backoff here or longer than second wait
        Requeue:      true,
    }, nil
```

Reply:
// RunInformersAndControllers the syncHandler, passing it the namespace/Name string of the
	// resource to be synced.
	if result, err := c.Do.Reconcile(req); err != nil {
		c.Queue.AddRateLimited(req)
		log.Error(err, "Reconciler error", "controller", c.Name, "request", req)
		ctrlmetrics.ReconcileErrors.WithLabelValues(c.Name).Inc()
		ctrlmetrics.ReconcileTotal.WithLabelValues(c.Name, "error").Inc()
		return false
	} else if result.RequeueAfter > 0 {
		c.Queue.AddAfter(req, result.RequeueAfter)
		ctrlmetrics.ReconcileTotal.WithLabelValues(c.Name, "requeue_after").Inc()
		return true
	} else if result.Requeue {
		c.Queue.AddRateLimited(req)
		ctrlmetrics.ReconcileTotal.WithLabelValues(c.Name, "requeue").Inc()
		return true
	}

The controller framework will automatically requeue the request if err is not nil.

Reply:
Missed that, thanks.